File: threedollar-api/src/main/java/com/depromeet/threedollar/api/service/visit/VisitHistoryService.java
package com.depromeet.threedollar.api.service.visit;

import com.depromeet.threedollar.domain.event.visit.VisitHistoryAddedEvent;
import com.depromeet.threedollar.api.service.store.StoreServiceUtils;
import com.depromeet.threedollar.api.service.visit.dto.request.AddVisitHistoryRequest;
import com.depromeet.threedollar.api.service.visit.dto.request.RetrieveMyVisitHistoriesRequest;
import com.depromeet.threedollar.api.service.visit.dto.response.VisitHistoriesScrollResponse;
import com.depromeet.threedollar.domain.collection.common.ScrollPaginationCollection;
import com.depromeet.threedollar.domain.domain.store.Store;
import com.depromeet.threedollar.domain.domain.store.StoreRepository;
import com.depromeet.threedollar.domain.domain.visit.VisitHistory;
import com.depromeet.threedollar.domain.domain.visit.VisitHistoryRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.context.ApplicationEventPublisher;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.List;

@RequiredArgsConstructor
@Service
public class VisitHistoryService {

    private final ApplicationEventPublisher eventPublisher;
    private final StoreRepository storeRepository;
    private final VisitHistoryRepository visitHistoryRepository;

    @Transactional
    public void addVisitHistory(AddVisitHistoryRequest request, Long userId, LocalDate dateOfVisit) {
        Store store = StoreServiceUtils.findStoreById(storeRepository, request.getStoreId());
        VisitHistoryServiceUtils.validateNotVisitedToday(visitHistoryRepository, request.getStoreId(), userId, dateOfVisit);
        VisitHistory visitHistory = visitHistoryRepository.save(request.toEntity(store, userId, dateOfVisit));
        eventPublisher.publishEvent(VisitHistoryAddedEvent.of(visitHistory.getId(), userId));
    }

    @Transactional(readOnly = true)
    public VisitHistoriesScrollResponse retrieveMyVisitHistories(RetrieveMyVisitHistoriesRequest request, Long userId) {
        List<VisitHistory> visitHistoriesWithNextCursor = visitHistoryRepository.findAllByUserIdWithScroll(userId, request.getCursor(), request.getSize() + 1);
        ScrollPaginationCollection<VisitHistory> visitHistoriesScroll = ScrollPaginationCollection.of(visitHistoriesWithNextCursor, request.getSize());
        return VisitHistoriesScrollResponse.of(visitHistoriesScroll);
    }

}

File: threedollar-api/src/main/java/com/depromeet/threedollar/api/service/visit/VisitHistoryService.kt
package com.depromeet.threedollar.api.service.visit

import com.depromeet.threedollar.api.controller.visit.dto.request.RetrieveVisitHistoryRequest
import com.depromeet.threedollar.api.service.store.StoreServiceUtils
import com.depromeet.threedollar.api.service.visit.dto.request.AddVisitHistoryRequest
import com.depromeet.threedollar.api.service.visit.dto.response.VisitHistoryResponse
import com.depromeet.threedollar.common.exception.model.ConflictException
import com.depromeet.threedollar.domain.domain.store.StoreRepository
import com.depromeet.threedollar.domain.domain.visit.VisitHistory
import com.depromeet.threedollar.domain.domain.visit.VisitHistoryRepository
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.time.LocalDate

@Service
class VisitHistoryService(
    private val storeRepository: StoreRepository,
    private val visitHistoryRepository: VisitHistoryRepository
) {

    @Transactional
    fun addStoreVisitHistory(request: AddVisitHistoryRequest, userId: Long) {
        StoreServiceUtils.validateExistsStore(storeRepository, request.storeId)
        val today = LocalDate.now()
        validateNotVisitedToday(visitHistoryRepository, request.storeId, userId, today)
        visitHistoryRepository.save<VisitHistory>(request.toEntity(userId, today))
    }

    @Transactional(readOnly = true)
    fun retrieveStoreVisitHistory(request: RetrieveVisitHistoryRequest): Map<LocalDate, List<VisitHistoryResponse>> {
        val today = LocalDate.now()
        val visitHistories =
            visitHistoryRepository.findVisitWithUserByStoreIdBetweenDate(request.storeId, today.minusWeeks(1), today)

        return visitHistories.asSequence()
            .map { VisitHistoryResponse.of(it) }
            .groupBy { history -> history.dateOfVisit }
            .toMap()
    }

}

fun validateNotVisitedToday(
    visitHistoryRepository: VisitHistoryRepository,
    storeId: Long?,
    userId: Long?,
    today: LocalDate?
) {
    if (visitHistoryRepository.existsByStoreIdAndUserIdAndDateOfVisit(storeId, userId, today)) {
        throw ConflictException(String.format("유저 (%s)는 (%s)에 (%s) 가게를 이미 방문 인증하였습니다", userId, today, storeId))
    }
}
